srk/src/weakTheoryInvariants.ml

```diff
@@ -0,0 +1,204 @@
+open Syntax
+open BatPervasives
+
+module TF = TransitionFormula
+module P = Polynomial
+module PC = PolynomialCone
+module WTS = WeakSolver
+module V = Linear.QQVector
+
+
+include Log.Make(struct let name = "WeakTheoryInvariants" end)
+
+let pp_dim srk = (fun formatter i ->
+    try Format.fprintf formatter "%a" (pp_symbol srk) (symbol_of_int i)
+    with _ -> Format.fprintf formatter "1")
+
+(* Do all symbols in a polynomial satisfy the given predicate? *)
+let for_all_vars predicate polynomial =
+  BatEnum.for_all
+    (fun (_, mono) ->
+       BatEnum.for_all
+         (fun (dim, _) -> predicate (symbol_of_int dim))
+         (P.Monomial.enum mono))
+    (P.QQXs.enum polynomial)
+
+(* Find polynomials that can be written as the sum of a linear term over a
+   designated set of variables and polynomial over the remaining variables. *)
+let filter_polys_linear_in_dims predicate polys =
+  BatList.filter_map
+    (fun poly ->
+       let lin, higher = P.QQXs.split_linear poly in
+       if for_all_vars (not % predicate) higher then
+         let (lin, higher) =
+           BatEnum.fold
+             (fun (lin, higher) (coeff, dim) ->
+                if dim == Linear.const_dim then
+                  (lin, P.QQXs.add_term coeff P.Monomial.one higher)
+                else if predicate (symbol_of_int dim) then
+                  (V.add_term coeff dim lin, higher)
+                else
+                  (lin, P.QQXs.add_term coeff (P.Monomial.singleton dim 1) higher))
+             (V.zero, higher)
+             (V.enum lin)
+         in
+         Some (lin, higher)
+       else
+         None)
+    polys
+
+let project_down_to_linear predicate pc =
+  let order =
+    P.Monomial.block [not % predicate % symbol_of_int] P.Monomial.degrevlex
+  in
+  let ideal =
+    PolynomialCone.get_ideal pc
+    |> P.Rewrite.reorder_groebner order
+    |> P.Rewrite.restrict (fun m ->
+        match P.Monomial.destruct_var m with
+        | Some v -> predicate (symbol_of_int v)
+        | None -> false)
+  in
+  PolynomialCone.make_cone ideal []
+
+let find_inv_functionals dx_dims implied_ideal =
+  let open Linear in
+  let basis = P.Rewrite.generators implied_ideal in
+  let polys = filter_polys_linear_in_dims dx_dims basis in
+  let lin_part_mat =
+    QQMatrix.transpose (QQMatrix.of_rows (BatList.map fst polys))
+  in
+  let const_parts =
+    BatList.mapi
+      (fun num ( _, const) -> (P.QQXs.coeff (P.Monomial.one) const, num))
+      polys
+  in
+  let const_mat = QQMatrix.of_rows [QQVector.of_list const_parts] in
+  let all_dims = BatList.of_enum (0 -- (BatList.length const_parts - 1)) in
+  let null_space = nullspace const_mat all_dims in
+  BatList.map
+    (fun base_vec -> QQMatrix.vector_right_mul lin_part_mat base_vec)
+    null_space
+
+let find_tf_invs srk tr_symbols loop_counter tf =
```

Comment:
Rename to something more descriptive.  Why does this take in tr_symbols *and* a transition formula?

Reply:
done